File: backend/src/main/java/dev/coldhands/pair/stairs/backend/usecase/BackendCombinationHistoryRepository.java
package dev.coldhands.pair.stairs.backend.usecase;

import dev.coldhands.pair.stairs.backend.infrastructure.mapper.CombinationMapper;
import dev.coldhands.pair.stairs.backend.infrastructure.persistance.entity.CombinationEventEntity;
import dev.coldhands.pair.stairs.backend.infrastructure.persistance.repository.CombinationEventRepository;
import dev.coldhands.pair.stairs.core.domain.Combination;
import dev.coldhands.pair.stairs.core.domain.CombinationHistoryRepository;
import dev.coldhands.pair.stairs.core.domain.pairstream.PairStream;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;

import java.util.List;

public class BackendCombinationHistoryRepository implements CombinationHistoryRepository<PairStream> {

    private final CombinationEventRepository repository;

    public BackendCombinationHistoryRepository(CombinationEventRepository repository) {
        this.repository = repository;
    }

    @Override
    public List<Combination<PairStream>> getMostRecentCombinations(int count) {
        final PageRequest pageRequest = PageRequest.of(0, count, Sort.by(Sort.Direction.DESC, "date"));

        final List<CombinationEventEntity> entities = repository.findAll(pageRequest).toList();

        return entities.stream()
                .map(CombinationEventEntity::getCombination)
                .map(CombinationMapper::entityToCore)
                .toList();
    }
}

File: backend/src/main/kotlin/dev/coldhands/pair/stairs/backend/usecase/BackendCombinationHistoryRepository.kt
package dev.coldhands.pair.stairs.backend.usecase

import dev.coldhands.pair.stairs.backend.domain.combination.CombinationEventDao
import dev.coldhands.pair.stairs.core.domain.Combination
import dev.coldhands.pair.stairs.core.domain.CombinationHistoryRepository
import dev.coldhands.pair.stairs.core.domain.pairstream.PairStream as CorePairStream

class BackendCombinationHistoryRepository(
    private val combinationEventDao: CombinationEventDao,
) : CombinationHistoryRepository<CorePairStream> {
    override fun getMostRecentCombinations(count: Int): List<Combination<CorePairStream>> {
        return combinationEventDao.getMostRecentCombinationEvents(count)
            .map { event ->
                Combination(event.combination.map {
                    val developerIdsAsString = it.developerIds
                        .map { developerId -> developerId.value.toString() }
                        .toSet()
                    CorePairStream(
                        developerIdsAsString,
                        it.streamId.value.toString()
                    )
                }
                    .toMutableSet()
                )
            }
    }
}